heatmap/src/heatmap.rs

```diff
@@ -297,166 +344,153 @@ impl Heatmap {
-                // note: we use parking_lot mutex as it will not be poisoned by
-                // a thread panic while locked.
-                if let Some(_lock) = self.lock.try_lock() {
-                    // now that we have the lock, check that we still need to
-                    // tick forward
-                    if time < self.next_tick.load(Ordering::Relaxed) {
-                        return;
-                    }
-
-                    // calculate the index of the next current slice
-                    let current = self.current.load(Ordering::Relaxed);
-                    let mut next = current + 1;
-                    if next >= self.slices.len() {
-                        next -= self.slices.len();
-                    }
+    fn idx_delta(&self, idx: usize, delta: i64) -> usize {
+        (idx + (self.slices() as i64 + delta) as usize) % self.slices()
+    }
 
-                    // move current and next_tick forward
-                    self.current.store(next, Ordering::Relaxed);
-                    self.next_tick.fetch_add(self.resolution, Ordering::Relaxed);
+    // compute the current slice index based on `tick_at` and `tick_origin`
+    fn slice_idx(&self, tick_at: Instant) -> usize {
+        let ntick =
+            tick_at.duration_since(self.tick_origin).as_nanos() / self.resolution.as_nanos();
+        (ntick - 1) as usize % self.slices()
+    }
 
-                    // now we have a slice to subtract and clear from the summary
-                    // this is the histogram that is one ahead of our new current
-                    // position
-                    let mut to_clear = next + 1;
+    // Internal function which handles all the housekeeping tasks that come due
+    // as the clock advances- primarily updating the time windows covered by each
+    // individual histogram, and cleaning up the values stored in buckets when
+    // the histogram is assigned to handle a new time slice.
+    //
+    // It returns the `tick_at` value which indicates the upper bound of the
+    // current `span`, the index of the most recent `Histogram` slice, and by
+    // how many ticks the heatmap moved forward.
+    fn tick(&self, now: Instant) -> (Instant, usize, usize) {
+        loop {
+            loop {
+                let tick_at = self.tick_at.load(Ordering::Relaxed);
+                // this is the common case when a heatmap is frequently updated, such as in a busy
+                // service.
+                if now < tick_at {
+                    return (tick_at, self.slice_idx(tick_at), 0);
+                }
 
-                    // check if we need to wrap around to the start
-                    if to_clear >= self.slices.len() {
-                        to_clear -= self.slices.len();
+                let ticks_forward =
+                    now.duration_since(tick_at).as_nanos() / self.resolution.as_nanos() + 1;
+                let mut new_tick = self.tick_at.load(Ordering::Relaxed);
```

Comment:
Is this load necessary?

Reply:
`tick_at` value could change, and if so the `compare_exchange` below would fail. In that regard using atomic-add and snapshot-and-add should be equivalent. But obtaining a local copy means we don't need to use atomic `fetch_add` within the `for` loop below this line.